### ISSUE TITLE
Issue with the domain of a function obtained with diff() method

### ISSUE THREAD

**Issue by CarmineVDV** [sympy/sympy#21046]
I've problems with the following code.

> from sympy import *
from IPython.display import Math
t = Symbol('t', real=True)
display(Math(latex("t is complex? " + str(t.is_complex))))
display(Math(latex("t is real? " + str(t.is_real))))
display(Math(latex("t is imaginary? " + str(t.is_imaginary))))
print()
f = Function('f', real = True)(t)
display(Math(latex("f is complex? " + str(f.is_complex))))
display(Math(latex("f is real? " + str(f.is_real))))
display(Math(latex("f is imaginary? " + str(f.is_imaginary))))
print()
g = f.diff(t)
display(Math("g(t)=" + latex(g)))
print()
display(Math(latex("g is complex? " + str(g.is_complex))))
display(Math(latex("g is real? " + str(g.is_real))))
display(Math(latex("g is imaginary? " + str(g.is_imaginary))))
print()
display(Math("Re\{g(t)\}=" + latex(re(g))))
display(Math("Im\{g(t)\}=" + latex(im(g))))

In output you can see that the function g(t) is the first derivative of the function f(t), but the code

> g.is_complex

returns None.
As before also 

> g.is_real
g.is_imaginary

return None.
The question is, defining t and f(t) as real, shouldn't g(t) be real?

> im(g)

shouldn't be 0?
However, shouldn't g(t) be complex anyway?

if it can be useful, i am using Python 3.7.1 and Sympy 1.7.1

**Comment by oscarbenjamin** [comment]
> The question is, defining t and f(t) as real, shouldn't g(t) be real?

Possibly.

It is also possible that `f(t)` is real but `f'(t)` doesn't exist or is "infinite". Maybe that case doesn't need to be considered though.

**Comment by CarmineVDV** [comment]
I agree with your considerations, but I just think that in any case, if g exists, then it should be real; for this reason I expected im{g(t)}=0.
The problem arose following the porting of a code written in maxima, in which the behavior was as expected.
In fact by executing in maxima the following code:

>declare(t, real);
declare(f, real);
f = depends(f, t);
're_f = re_f : realpart(f);
'im_f = im_f : imagpart(f);
'g = g : diff(f, t);
're_g = re_g : realpart(g);
'im_g = im_g : imagpart(g);

everything works.
Below is the output:

> done
done
f=[f(t)]
re_f=f
im_f=0
g='diff(f,t,1)
re_g='diff(f,t,1)
im_g=0

**Comment by oscarbenjamin** [comment]
Maybe it would make sense for `f'(t)` to have `is_extended_real` as True which would allow `oo` and `-oo`. Then `re` and `im` would work.
